### ISSUE TITLE
A rule to prefer SAFE_CAST over CAST in BigQuery

### ISSUE THREAD

**Issue by kitsunde** [sqlfluff/sqlfluff#4000]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

In BQ there is both `CAST` and `SAFE_CAST` where `CAST` errors when the value can't be type cast and `SAFE_CAST` returns a `NULL`.

Ref: https://cloud.google.com/bigquery/docs/reference/standard-sql/functions-and-operators#safe_casting

### Use case

This should fail:

```sql
SELECT CAST('not a number' AS FLOAT64)
```

This should not:

```sql
SELECT SAFE_CAST('not a number' AS FLOAT64)
```

### Dialect

BigQuery

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by greg-finley** [comment]
I'm torn. There's some valid use case to know that you are not casting values properly vs silently ignoring them. Is it necessarily better to nudge people to the latter?

**Comment by kitsunde** [comment]
In our case we aren’t querying our own data hat can be controlled beforehand, but third party data where we treat invalid data as NULL in aggregation.

I’m not suggesting a rule that needs to be always enabled just available.

My concern is that I don’t think it’s realistic long term that people making changes and reviewing will be aware that SAFE_CAST even exists, and that CAST is really UNSAFE_CAST because it looks innocent and accurate.

**Comment by barrywhart** [comment]
I would definitely want this rule to be disabled by default. Too dangerous as a general recommendation. Note that few if any SQLFluff rules change the behavior of existing code.

**Comment by tunetheweb** [comment]
This can easily be configured using the existing rule [CV09](https://docs.sqlfluff.com/en/stable/rules.html#sqlfluff.rules.sphinx.Rule_CV09) (formally known as L062).

That doesn't address the question of whether there should be a more generic rule (to try to encourage this use).

My main concern is that this changes behaviour and silently allows a badly written query to continue when perhaps it shouldn't, and perhaps silently return incorrect results. I'd argue it's better to throw the error in these cases and let the SQL developer explicitly add the `SAFE_` prefix only if they are OK with ignoring it in that case.

So not sure this is a good rule for a linter to promote, therefore am probably against adding this as a rule. But CV09 allows you to configure that if it's something you do want to enforce in your code base.

**Comment by peterbud** [comment]
Closing as having a workaround in place.
